[refcycle-logger] Handle ReferenceErrors

czardoz opened the pull request:
The ref-cycle logger currently throws errors if hit with a weakref proxy object, e.g:

```
class Klass:
    pass

obj = Klass()
proxy = weakref.proxy(obj)
del obj

# This should raise ReferenceError
isinstance(proxy, type)
```

This situation happens for the ref-cycle logger if a proxy object ends up in `gc.garbage` after `gc.collect()` is called. There's not much we can do when we hit this situation (since the underlying object has been GC'd already). Just swallow the exception and proceed with the rest of the objects.

Differential Revision: D88090361

cyyever commented on torch/utils/viz/_cycles.py:
```diff
@@ -212,7 +212,11 @@ def object_annotation(obj):
-        body = ','.join(repr(x) if isinstance(x, BASE_TYPES) else type(x).__name__ for x in obj[:8])
+        def safe_repr(x):
+            if isinstance(x, weakref.ProxyTypes):
+                return "weakref (dead)"
```
Add `repr(x)` to the weakref string?

czardoz replied:
Trying to do anything with the proxy after the object is freed is a `ReferenceError`.

czardoz committed:
[refcycle-logger] Handle ReferenceErrors (#169315)

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/169315

Differential Revision: D88090361
